File: scripts/map-assets/build-unified-satellite.mjs
// T-090.1.2.8 — Build the unified satellite bundle (tbd-sat v1) from a full-extent ortho.
//
// One asset per terrain satellite: a GLB-style container holding the FULL GPU mip chain
// (base → 1×1, dims = max(1, floor(prev/2)) — the GL mip rule) as lossless WebP (VP8L)
// blocks. The frontend fetches it ONCE, decodes each mip via createImageBitmap, and uploads
// them into a single luma.gl texture — no per-viewport tile HTTP, no BitmapLayer pop-in
// (the T-090.1.2.8 "Reforger map feel" contract). Replaces the 5461-tile pyramid as the
// primary delivery; the pyramid stays on disk as the manifest-flagged fallback.
//
// Container layout (offsets absolute from file start, little-endian):
//   bytes 0..3   magic  ASCII "TBDS"
//   bytes 4..7   u32    formatVersion = 1
//   bytes 8..11  u32    jsonLength
//   bytes 12..   UTF-8 JSON index (mips[].tiles[] with offset/length per VP8L block)
//   then         concatenated WebP blocks
//
// A mip level splits into ceil(dim/tileThreshold) tiles per axis (Everon: level 0 12800² →
// 2×2 of 6400², everything else 1×1). Tiling keeps every block under the VP8L 16383 px cap
// for future terrains, parallelizes the slow lossless encodes across cores at build time,
// and lets the browser decode the base level in parallel workers. Tile x/y are pixel
// offsets in IMAGE space (row 0 = top = north — same orientation contract as the pyramid
// input: no flips anywhere on this path).
//
// Usage:
//   node scripts/map-assets/build-unified-satellite.mjs \
//     --input packages/map-assets/everon/staging/sap/everon-sap-ortho.png \
//     --out   packages/map-assets/everon/satellite/everon-sat.tbd-sat \
//     --terrain everon [--tile-threshold 8192]
//
// Requires magick (ImageMagick 7) + cwebp on PATH (same toolchain as build-tile-pyramid.sh).
import { createHash } from 'node:crypto';
import { execFileSync, execFile } from 'node:child_process';
import { mkdtempSync, rmSync, mkdirSync, readFileSync, writeFileSync, existsSync } from 'node:fs';
import { tmpdir, cpus } from 'node:os';
import { join, dirname, resolve } from 'node:path';
import { promisify } from 'node:util';

const execFileP = promisify(execFile);

function arg(name, fallback) {
  const i = process.argv.indexOf(`--${name}`);
  return i >= 0 ? process.argv[i + 1] : fallback;
}

const INPUT = arg('input');
const OUT = arg('out');
const TERRAIN = arg('terrain', 'everon');
const TILE_THRESHOLD = Number(arg('tile-threshold', '8192'));

if (!INPUT || !OUT) {
  console.error(
    'Usage: node build-unified-satellite.mjs --input <ortho.png> --out <bundle.tbd-sat> --terrain everon [--tile-threshold 8192]',
  );
  process.exit(2);
}
if (!existsSync(INPUT)) {
  console.error(`input not found: ${INPUT}`);
  process.exit(1);
}
for (const tool of ['magick', 'cwebp']) {
  try {
    execFileSync(tool, ['-version'], { stdio: 'ignore' });
  } catch {
    console.error(`${tool} required on PATH`);
    process.exit(1);
  }
}

const MAGIC = 'TBDS';
const FORMAT_VERSION = 1;

// Terrain contract mirrors packages/map-assets/<terrain>/manifest.json worldBounds.
const WORLD_BOUNDS = {
  everon: [0, 0, 12800, 12800],
  arland: [0, 0, 4096, 4096],
};
const worldBounds = WORLD_BOUNDS[TERRAIN];
if (!worldBounds) {
  console.error(`unknown terrain "${TERRAIN}" (add its worldBounds to build-unified-satellite.mjs)`);
  process.exit(1);
}

const work = mkdtempSync(join(tmpdir(), 'tbd-sat-'));
process.on('exit', () => rmSync(work, { recursive: true, force: true }));

const t0 = Date.now();
const log = (m) => console.log(`[tbd-sat] ${m}`);

// --- 1. Normalize: strip alpha + metadata, force sRGB (same normalize as the pyramid). ---
const norm = join(work, 'mip0.png');
log(`normalizing ${INPUT}`);
execFileSync('magick', [INPUT, '-alpha', 'off', '-colorspace', 'sRGB', norm]);
const [srcW, srcH] = execFileSync('magick', ['identify', '-format', '%w %h', norm])
  .toString()
  .trim()
  .split(' ')
  .map(Number);
log(`source ${srcW}x${srcH}; tileThreshold=${TILE_THRESHOLD}`);

// --- 2. Mip chain dims: base → 1×1 with the GL rule so mipCount always equals
//        floor(log2(max(w,h)))+1 and lines up with luma.gl getMipLevelCount. ---
const levels = [];
for (let w = srcW, h = srcH; ; w = Math.max(1, Math.floor(w / 2)), h = Math.max(1, Math.floor(h / 2))) {
  levels.push({ width: w, height: h });
  if (w === 1 && h === 1) break;
}
log(`mip chain: ${levels.length} levels (${srcW} → 1)`);

// --- 3. Per level: cascade-halve from the previous level PNG (standard mip practice — one
//        full-res Lanczos pass then successive halvings), crop over-threshold levels into a
//        tile grid, and encode every tile as lossless WebP. Encodes are fanned out across
//        cores; level-0 quadrants are what parallelize the dominant cost. ---
const encodeJobs = []; // {level, gridX, gridY, png, webp}
for (let level = 0; level < levels.length; level++) {
  const { width, height } = levels[level];
  const lvPng = join(work, `mip${level}.png`);
  if (level > 0) {
    execFileSync('magick', [join(work, `mip${level - 1}.png`), '-resize', `${width}x${height}!`, lvPng]);
  }
  const cols = Math.ceil(width / TILE_THRESHOLD);
  const rows = Math.ceil(height / TILE_THRESHOLD);
  const tileW = Math.ceil(width / cols);
  const tileH = Math.ceil(height / rows);
  const lvDir = join(work, `level-${level}`);
  mkdirSync(lvDir);
  if (cols === 1 && rows === 1) {
    encodeJobs.push({ level, gridX: 0, gridY: 0, w: width, h: height, png: lvPng, webp: join(lvDir, 't0-0.webp') });
  } else {
    // magick -crop WxH +adjoin slices row-major: index i → col i%cols, row i/cols.
    execFileSync('magick', [lvPng, '-crop', `${tileW}x${tileH}`, '+repage', '+adjoin', join(lvDir, 'tile_%d.png')]);
    for (let gy = 0; gy < rows; gy++) {
      for (let gx = 0; gx < cols; gx++) {
        const w = Math.min(tileW, width - gx * tileW);
        const h = Math.min(tileH, height - gy * tileH);
        encodeJobs.push({
          level,
          gridX: gx,
          gridY: gy,
          w,
          h,
          png: join(lvDir, `tile_${gy * cols + gx}.png`),
          webp: join(lvDir, `t${gx}-${gy}.webp`),
        });
      }
    }
  }
  levels[level].cols = cols;
  levels[level].rows = rows;
  levels[level].tileW = tileW;
  levels[level].tileH = tileH;
}

const parallel = Math.max(1, cpus().length - 1);
log(`encoding ${encodeJobs.length} VP8L blocks (parallel=${parallel})`);
{
  const queue = [...encodeJobs];
  const worker = async () => {
    for (let job = queue.shift(); job; job = queue.shift()) {
      await execFileP('cwebp', ['-quiet', '-lossless', '-mt', job.png, '-o', job.webp]);
    }
  };
  await Promise.all(Array.from({ length: parallel }, worker));
}

// --- 4. Assemble container. Compute the JSON with offset placeholders sized by a two-pass
//        layout: payload offsets depend on jsonLength, so lay blocks out relative to the
//        payload start first, then rewrite as absolute once the JSON byte length is fixed
//        (offsets grow the JSON, so iterate until stable — bounded, tiny). ---
const inputSha256 = createHash('sha256').update(readFileSync(INPUT)).digest('hex');
let sourceMeta;
const metaPath = join(dirname(INPUT), 'TBD_SatExport_meta.json');
if (existsSync(metaPath)) {
  const m = JSON.parse(readFileSync(metaPath, 'utf8'));
  sourceMeta = { source: m.source, seamRepair: m.seamRepair, generatedAt: m.generatedAt };
}

const blocks = encodeJobs.map((j) => ({ job: j, buf: readFileSync(j.webp) }));
const mips = levels.map((lv, level) => ({
  level,
  width: lv.width,
  height: lv.height,
  tiles: blocks
    .filter((b) => b.job.level === level)
    .sort((a, b) => a.job.gridY - b.job.gridY || a.job.gridX - b.job.gridX)
    .map((b) => ({
      x: b.job.gridX * lv.tileW,
      y: b.job.gridY * lv.tileH,
      width: b.job.w,
      height: b.job.h,
      offset: 0, // patched below
      length: b.buf.length,
    })),
}));

const index = {
  formatVersion: FORMAT_VERSION,
  terrainId: TERRAIN,
  worldBounds,
  metersPerPixel: worldBounds[2] / srcW,
  source: sourceMeta?.source ?? 'unknown',
  sourceMeta,
  encoding: 'webp-lossless',
  createdAt: new Date().toISOString(),
  inputSha256,
  baseWidthPx: srcW,
  baseHeightPx: srcH,
  mipCount: levels.length,
  mips,
};

// Iterate offset patching until the JSON length stops moving (number widths stabilize).
const orderedTiles = mips.flatMap((m) => m.tiles);
let jsonBuf;
for (let jsonLen = 0; ; ) {
  let offset = 12 + jsonLen;
  for (const t of orderedTiles) {
    t.offset = offset;
    offset += t.length;
  }
  jsonBuf = Buffer.from(JSON.stringify(index), 'utf8');
  if (jsonBuf.length === jsonLen) break;
  jsonLen = jsonBuf.length;
}

const header = Buffer.alloc(12);
header.write(MAGIC, 0, 'ascii');
header.writeUInt32LE(FORMAT_VERSION, 4);
header.writeUInt32LE(jsonBuf.length, 8);

mkdirSync(dirname(resolve(OUT)), { recursive: true });
writeFileSync(OUT, Buffer.concat([header, jsonBuf, ...blocks.map((b) => b.buf)]));

const total = 12 + jsonBuf.length + blocks.reduce((s, b) => s + b.buf.length, 0);
for (const m of mips) {
  const bytes = m.tiles.reduce((s, t) => s + t.length, 0);
  log(`  level ${String(m.level).padStart(2)}  ${String(m.width).padStart(5)}px  ${m.tiles.length} block(s)  ${(bytes / 1e6).toFixed(2)} MB`);
}
log(`wrote ${OUT}  ${(total / 1e6).toFixed(1)} MB in ${((Date.now() - t0) / 1000).toFixed(0)}s`);
log(`manifest block:`);
console.log(
  JSON.stringify(
    {
      delivery: 'unified',
      unified: {
        path: `satellite/${OUT.split('/').pop()}`,
        url: `/map-assets/${TERRAIN}/satellite/${OUT.split('/').pop()}`,
        encoding: 'tbd-sat-v1',
        baseWidthPx: srcW,
        baseHeightPx: srcH,
        mipCount: levels.length,
        bytes: total,
      },
    },
    null,
    2,
  ),
);

File: tools_v2/developer-tools/src/map_raster_pipeline/satellite_archive/build_unified_satellite.rs
use super::*;

pub fn build_unified_satellite(
    input: &Path,
    out: &Path,
    terrain: &str,
    tile_threshold: usize,
    container_version: u16,
) -> Result<u8> {
    use sha2::Digest as _;
    if container_version != 1 && container_version != 2 {
        eprintln!("unsupported --container-version {container_version} (expected 1 or 2)");
        return Ok(1);
    }
    let world_bounds: [u64; 4] = match terrain {
        "everon" => [0, 0, 12800, 12800],
        "arland" => [0, 0, 4096, 4096],
        _ => {
            eprintln!("unknown terrain \"{terrain}\" (add its worldBounds to the unified builder)");
            return Ok(1);
        }
    };
    if !input.exists() {
        eprintln!("input not found: {}", input.display());
        return Ok(1);
    }
    let t0 = std::time::Instant::now();
    let log = |m: &str| println!("[tbd-sat] {m}");

    log(&format!("normalizing {}", input.display()));
    let base = image_operations::load_png_rgb(input)?;
    let (src_w, src_h) = (base.w, base.h);
    log(&format!(
        "source {src_w}x{src_h}; tileThreshold={tile_threshold}"
    ));

    // Mip chain dims: base → 1×1 with the GL rule.
    let dims = mip_dims(src_w, src_h);
    log(&format!("mip chain: {} levels ({src_w} → 1)", dims.len()));

    // Cascade-halve + tile + encode (rayon-free: encode sequentially — image-webp lossless
    // is fast enough for the rebuild-smoke acceptance; parallelism can come later).
    let mut blocks: Vec<TileBuf> = Vec::new();
    let mut level_meta = Vec::new();
    let mut current = base;
    for (level, &(lw, lh)) in dims.iter().enumerate() {
        if level > 0 {
            current = image_operations::resize_rgb(&current, lw, lh);
        }
        let cols = lw.div_ceil(tile_threshold);
        let rows = lh.div_ceil(tile_threshold);
        let tile_w = lw.div_ceil(cols);
        let tile_h = lh.div_ceil(rows);
        for gy in 0..rows {
            for gx in 0..cols {
                let x = gx * tile_w;
                let y = gy * tile_h;
                let tw = tile_w.min(lw - x);
                let th = tile_h.min(lh - y);
                let tile = if cols == 1 && rows == 1 {
                    current.data.clone()
                } else {
                    image_operations::crop_rgb(&current, x, y, tw, th)?.data
                };
                let buf = image_operations::encode_webp_lossless_rgb(&image_operations::Rgb8 {
                    w: tw,
                    h: th,
                    data: tile,
                })?;
                blocks.push(TileBuf {
                    level,
                    x,
                    y,
                    w: tw,
                    h: th,
                    buf,
                });
            }
        }
        level_meta.push((lw, lh));
    }
    log(&format!("encoded {} VP8L blocks", blocks.len()));

    let input_sha256 = {
        let mut hsh = sha2::Sha256::new();
        hsh.update(std::fs::read(input)?);
        hsh.finalize()
            .iter()
            .map(|b| format!("{b:02x}"))
            .collect::<String>()
    };
    let mut source_meta = Value::Null;
    let meta_path = input
        .parent()
        .unwrap_or(Path::new("."))
        .join("TBD_SatExport_meta.json");
    if meta_path.exists() {
        let m: Value = serde_json::from_str(&std::fs::read_to_string(&meta_path)?)?;
        let mut sm = serde_json::Map::new();
        sm.insert("source".into(), m["source"].clone());
        sm.insert("seamRepair".into(), m["seamRepair"].clone());
        sm.insert("generatedAt".into(), m["generatedAt"].clone());
        if m["waterComposite"].is_object() {
            let wc = &m["waterComposite"];
            sm.insert(
                "waterComposite".into(),
                json!({
                    "lane": wc["lane"], "oceanMaskSource": wc["oceanMaskSource"],
                    "inlandMaskSource": wc["inlandMaskSource"], "generatedAt": wc["generatedAt"],
                }),
            );
        }
        source_meta = Value::Object(sm);
    }

    let file = if container_version == 2 {
        let index = tbds_v2_index(&blocks, &level_meta, (src_w, src_h), tile_threshold)?;
        tbds_v2_bytes(&index, &blocks)?
    } else {
        build_tbds_v1_bytes(
            &blocks,
            &level_meta,
            (src_w, src_h),
            terrain,
            world_bounds,
            &source_meta,
            &input_sha256,
        )?
    };
    if let Some(parent) = out.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(out, &file)?;

    // Driven by the encoded blocks rather than by either index, so the receipt says the same thing
    // for both container versions.
    for (level, &(lw, _)) in level_meta.iter().enumerate() {
        let (n, bytes) = blocks
            .iter()
            .filter(|b| b.level == level)
            .fold((0usize, 0u64), |(n, bytes), b| {
                (n + 1, bytes + b.buf.len() as u64)
            });
        log(&format!(
            "  level {level:>2}  {lw:>5}px  {n} block(s)  {:.2} MB",
            bytes as f64 / 1e6
        ));
    }
    log(&format!(
        "wrote {}  {:.1} MB in {:.0}s (tbd-sat v{container_version})",
        out.display(),
        file.len() as f64 / 1e6,
        t0.elapsed().as_secs_f64()
    ));
    log("manifest block:");
    let base_name = out.file_name().unwrap_or_default().to_string_lossy();
    println!(
        "{}",
        serde_json::to_string_pretty(&json!({
            "delivery": "unified",
            "unified": {
                "path": format!("satellite/{base_name}"),
                "url": format!("/map-assets/{terrain}/satellite/{base_name}"),
                "encoding": format!("tbd-sat-v{container_version}"),
                "baseWidthPx": src_w,
                "baseHeightPx": src_h,
                "mipCount": dims.len(),
                "bytes": file.len(),
            },
        }))?
    );
    Ok(0)
}

/// The v1 container bytes: `"TBDS"`, formatVersion 1, jsonLength, the hand-packed JSON table, then
/// the payload. Retained verbatim behind `--container-version 1` because `everon-sat.tbd-sat` is
/// committed in this shape and stays that way until T-935.13 regenerates it.
pub(crate) fn build_tbds_v1_bytes(
    blocks: &[TileBuf],
    level_meta: &[(usize, usize)],
    base: (usize, usize),
    terrain: &str,
    world_bounds: [u64; 4],
    source_meta: &Value,
    input_sha256: &str,
) -> Result<Vec<u8>> {
    let (src_w, src_h) = base;
    let mut mips: Vec<Value> = Vec::new();
    for (level, &(lw, lh)) in level_meta.iter().enumerate() {
        let tiles: Vec<Value> = blocks
            .iter()
            .filter(|b| b.level == level)
            .map(|b| {
                json!({ "x": b.x, "y": b.y, "width": b.w, "height": b.h, "offset": 0, "length": b.buf.len() })
            })
            .collect();
        mips.push(json!({ "level": level, "width": lw, "height": lh, "tiles": tiles }));
    }
    let mut index = json!({
        "formatVersion": 1,
        "terrainId": terrain,
        "worldBounds": world_bounds,
        "metersPerPixel": js_num(world_bounds[2] as f64 / src_w as f64),
        "source": if source_meta["source"].is_string() { source_meta["source"].clone() } else { json!("unknown") },
        "sourceMeta": source_meta,
        "encoding": "webp-lossless",
        "createdAt": iso_from_system_time(std::time::SystemTime::now()),
        "inputSha256": input_sha256,
        "baseWidthPx": src_w,
        "baseHeightPx": src_h,
        "mipCount": level_meta.len(),
        "mips": mips,
    });

    // Two-pass offset patch until the JSON length stabilizes.
    let mut json_buf: Vec<u8>;
    let mut json_len = 0usize;
    loop {
        let mut offset = 12 + json_len;
        for mip in index["mips"].as_array_mut().unwrap() {
            for t in mip["tiles"].as_array_mut().unwrap() {
                t["offset"] = json!(offset);
                offset += t["length"].as_u64().unwrap() as usize;
            }
        }
        json_buf = serde_json::to_string(&index)?.into_bytes();
        if json_buf.len() == json_len {
            break;
        }
        json_len = json_buf.len();
    }

    let mut file = Vec::with_capacity(12 + json_buf.len());
    file.extend_from_slice(b"TBDS");
    file.extend_from_slice(&1u32.to_le_bytes());
    file.extend_from_slice(&(json_buf.len() as u32).to_le_bytes());
    file.extend_from_slice(&json_buf);
    for b in blocks {
        file.extend_from_slice(&b.buf);
    }
    Ok(file)
}
